### DISCUSSION TITLE
bucket name does not match

### DISCUSSION THREAD

**Discussion by kendn1993** [minio/minio#13398] [Q&A] [answered]
## Current Behavior

Bucketname not match occurs occasionally during bucket building
![image](https://user-images.githubusercontent.com/27141040/136653910-e1e72e3a-92ae-4d8c-b2f9-5225c742e6b1.png)

## Steps to Reproduce (for bugs)
here is my code 
if (!client.bucketExists(BucketExistsArgs.builder().bucket(bucketName).build())) {
            client.makeBucket(MakeBucketArgs.builder().bucket(bucketName).build());
            setBucketPolicy(bucketName, READ_WRITE);
            //            client.setBucketPolicy(minioProp.getBucketName(), "*.*", PolicyType.READ_ONLY);
        }

## Your Environment

* Version used (`minio --version`):
minio version RELEASE.2021-10-08T23-58-24Z
* Server setup and configuration:
* Operating System and version (`uname -a`):
* Linux hadoop003 3.10.0-957.el7.x86_64 minio/minio#1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
This is not a server exception it looks like client exception, moving this to minio-java. Please share the entire code here @kendn1993

**Comment by balamurugana** [comment]
As per the trace, the server is returning S3 error. Nothing can be done at client side. Please enable `MinioClient.traceOn()` and open respective issue at server side with the output.

**Comment by harshavardhana** [comment]
> As per the trace, the server is returning S3 error. Nothing can be done at client side. Please enable `MinioClient.traceOn()` and open respective issue at server side with the output.

@kendn1993 please share this output here

**Comment by harshavardhana** [comment]
I check the server code there is no place where the server would return "bucket name does not match"

**Reply by kendn1993** [reply]
log shows the server return

**Comment by kendn1993** [comment]
---------START-HTTP---------
GET /test111111?location= HTTP/1.1
Host: 106.112.143.254:8333
Accept-Encoding: identity
User-Agent: MinIO (Windows 10; amd64) minio-java/8.3.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20211011T074035Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20211011/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*

HTTP/1.1 404
Server: nginx
Date: Mon, 11 Oct 2021 07:40:35 GMT
Content-Type: application/xml
Content-Length: 292
Connection: keep-alive
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: 16ACE9FC8911F4B3
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><BucketName>test111111</BucketName><Resource>/test111111</Resource><RequestId>16ACE9FC8911F4B3</RequestId><HostId>5ce445aa-1c3b-48ab-a35f-457b2e1cf45c</HostId></Error>
----------END-HTTP----------
---------START-HTTP---------
PUT /test111111 HTTP/1.1
Host: 106.112.143.254:8333
Accept-Encoding: identity
User-Agent: MinIO (Windows 10; amd64) minio-java/8.3.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20211011T074035Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20211011/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*

HTTP/1.1 200
Server: nginx
Date: Mon, 11 Oct 2021 07:40:35 GMT
Content-Length: 0
Connection: keep-alive
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Location: /test111111
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: 16ACE9FC8C0A9636
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block


----------END-HTTP----------
---------START-HTTP---------
PUT /test111111?policy= HTTP/1.1
Content-Type: application/json
Host: 106.112.143.254:8333
Accept-Encoding: identity
User-Agent: MinIO (Windows 10; amd64) minio-java/8.3.0
Content-MD5: FpQaoExhctzvXuLNTuxJwg==
x-amz-content-sha256: d7458a6ad7ce17fd62468303fd1bed9b87fa400bdf99259c9c6535b604fee68a
x-amz-date: 20211011T074035Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20211011/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*

{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["*"]},"Action":["s3:GetBucketLocation","s3:ListBucket","s3:ListBucketMultipartUploads"],"Resource":["arn:aws:s3:::$;{bucket}"]},{"Effect":"Allow","Principal":{"AWS":["*"]},"Action":["s3:DeleteObject","s3:GetObject","s3:ListMultipartUploadParts","s3:PutObject","s3:AbortMultipartUpload"],"Resource":["arn:aws:s3:::test111111/*"]}]}
HTTP/1.1 400
Server: nginx
Date: Mon, 11 Oct 2021 07:40:35 GMT
Content-Type: application/xml
Content-Length: 286
Connection: keep-alive
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Origin
Vary: Accept-Encoding
X-Amz-Request-Id: 16ACE9FC87F7665D
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>MalformedPolicy</Code><Message>bucket name does not match</Message><BucketName>test111111</BucketName><Resource>/test111111</Resource><RequestId>16ACE9FC87F7665D</RequestId><HostId>5ce445aa-1c3b-48ab-a35f-457b2e1cf45c</HostId></Error>
----------END-HTTP----------

**Comment by harshavardhana** [comment]
Yes because your policy is wrong @kendn1993 

```
{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":[""]},"Action":["s3:GetBucketLocation","s3:ListBucket","s3:ListBucketMultipartUploads"],"Resource":["arn:aws:s3:::$;{bucket}"]},{"Effect":"Allow","Principal":{"AWS":[""]},"Action":["s3:DeleteObject","s3:GetObject","s3:ListMultipartUploadParts","s3:PutObject","s3:AbortMultipartUpload"],"Resource":["arn:aws:s3:::test111111/*"]}]}
```

You cannot have `${bucket}` - that doesn't make any sense.

**Reply by kendn1993** [reply]
thanks, the problem is solved

**Comment by harshavardhana** [comment, marked as answer]
> here is my code
> if (!client.bucketExists(BucketExistsArgs.builder().bucket(bucketName).build())) {
> client.makeBucket(MakeBucketArgs.builder().bucket(bucketName).build());
> setBucketPolicy(bucketName, READ_WRITE);
> // client.setBucketPolicy(minioProp.getBucketName(), "_._", PolicyType.READ_ONLY);
> }

Also the API has changed you need to upgrade to latest minio-java version @kendn1993 

```java
      StringBuilder builder = new StringBuilder();
      builder.append("{\n");
      builder.append("    \"Statement\": [\n");
      builder.append("        {\n");
      builder.append("            \"Action\": [\n");
      builder.append("                \"s3:GetBucketLocation\",\n");
      builder.append("                \"s3:ListBucket\"\n");
      builder.append("            ],\n");
      builder.append("            \"Effect\": \"Allow\",\n");
      builder.append("            \"Principal\": \"*\",\n");
      builder.append("            \"Resource\": \"arn:aws:s3:::my-bucketname\"\n");
      builder.append("        },\n");
      builder.append("        {\n");
      builder.append("            \"Action\": \"s3:GetObject\",\n");
      builder.append("            \"Effect\": \"Allow\",\n");
      builder.append("            \"Principal\": \"*\",\n");
      builder.append("            \"Resource\": \"arn:aws:s3:::my-bucketname/myobject*\"\n");
      builder.append("        }\n");
      builder.append("    ],\n");
      builder.append("    \"Version\": \"2012-10-17\"\n");
      builder.append("}\n");
      minioClient.setBucketPolicy(
          SetBucketPolicyArgs.builder().bucket("my-bucketname").config(builder.toString()).build());
```

**Comment by avDec25** [comment]
String format template to set public Bucket policy.

```java
private final String PUBLIC_POLICY = "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":[\"*\"]},\"Action\":[\"s3:GetBucketLocation\",\"s3:ListBucket\",\"s3:ListBucketMultipartUploads\"],\"Resource\":[\"arn:aws:s3:::%s\"]},{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":[\"*\"]},\"Action\":[\"s3:PutObject\",\"s3:AbortMultipartUpload\",\"s3:DeleteObject\",\"s3:GetObject\",\"s3:ListMultipartUploadParts\"],\"Resource\":[\"arn:aws:s3:::%s/*\"]}]}";

minioClient.makeBucket(MakeBucketArgs.builder().bucket(bktName).build());
minioClient.setBucketPolicy(
                SetBucketPolicyArgs.builder()
                .bucket(bktName)
                .config(
                    String.format(PUBLIC_POLICY, bktName, bktName)
                )
                .build()
);
```
